pilot/pkg/config/kube/gateway/testdata/mesh.yaml

```diff
@@ -31,24 +31,8 @@ metadata:
   namespace: default
 spec:
   parentRefs:
-  - kind: Mesh
-    name: istio
-  hostnames: ["echo.default.svc.cluster.local"]
-  rules:
-  - backendRefs:
-    - name: echo
-      port: 80
----
-apiVersion: gateway.networking.k8s.io/v1alpha2
-kind: TLSRoute
-metadata:
-  name: echo
-  namespace: default
-spec:
-  parentRefs:
-  - kind: Mesh
-    name: istio
-  hostnames: ["some-sni.com"]
+  - kind: Service
```

Comment:
can we have a test with 2 routes for the same service? They should merge

Reply:
Done, but I noticed that there doesn't seem to be a way to prioritize merged rules, so the test case generates a merged virtual service as expected but the order isn't what you'd want:
```
  - route:
    - destination:
        host: echo.default.svc.domain.suffix
        port:
          number: 80
    - headers:
       request:
         add:
           my-added-header: added-value
    match:
    - uri:
        prefix: /path
    route:
    - destination:
        host: echo.default.svc.domain.suffix
        port:
          number: 80
```

Reply:
https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteRule defines how they should merge; this looks like we aren't properly following it.

For Gateway we have some logic to do this afaik - `sortHTTPRoutes`

Reply:
`sortHTTPRoutes` was being called, but the check for route-with-match vs route-without-match wasn't correct. I fixed that so it's working properly for that case now (see line 455 in conversion.go).

But, I'm not convinced that the rest of the checking really makes much sense. For example uri lenth comparison where one might be a regex vs another that is a prefix. Seems we might reorder things the user had explicitly ordered correctly in their rule. Also seems like we are only doing sorting with match[0], so what should happen if there are more than one?

Reply:
> the user had explicitly ordered correctly in their rule

HTTPRoute is an unordered list, so this sentence doesn't make sense in the context of the spec as written today. Your concerns are valid but we should either change the spec upstream or follow it

> Also seems like we are only doing sorting with match[0], so what should happen if there are more than one?

I think we split the matches so there is always only 1? If not, may be a bug